Implement AppManager::getAppPath

VicDeo opened the pull request:
## Description
- Allows to get the app path and app web path
- Kill legacy code that does the same. `OC_App::getAppPath` and `OC_App::getAppWebPath` are cut down to work via AppManager implementation
- It is possible to use some AppManager methods without Db connection now (e.g. when OC is not installed)

## Related Issue
https://github.com/owncloud/core/issues/25336

## Motivation and Context
- public API enhancement
- unit testing is possible now
- killing old crap

## How Has This Been Tested?
`\OC::$server->getAppManager()->getAppPath('customgroups');`
`\OC::$server->getAppManager()->getAppWebPath('customgroups');`

where 'customgroups' is your favorite app. In a various configs including multiple app directories.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

VicDeo committed:
Refactor getAppPath

phil-davis commented on lib/private/App/AppManager.php:
```diff
@@ -483,4 +492,77 @@ public function canInstall() {
+	}
+
+	/**
+	 * Get the directory for the given app.
```
Method description needs an update. (copy-pasted from method above but not yet modified)

PVince81 commented on lib/private/App/AppManager.php:
```diff
@@ -483,4 +492,97 @@ public function canInstall() {
+		if (trim($appId) === '') {
+			return false;
+		}
+		if (($appRoot = $this->findAppInDirectories($appId)) != false) {
```
feel free to leave it if it's just copy-pasted bad old code

PVince81 commented on lib/private/App/AppManager.php:
```diff
@@ -483,4 +492,97 @@ public function canInstall() {
+	 * @param string $appId
+	 * @return false|string[]
+	 */
+	protected  function findAppInDirectories($appId) {
```
double space between `protected` and `function`

PVince81 commented on lib/private/App/AppManager.php:
```diff
@@ -483,4 +492,97 @@ public function canInstall() {
+			$versionToLoad = [];
+			foreach ($possibleAppRoots as $possibleAppRoot) {
+				$version = $this->getAppVersionByPath($possibleAppRoot['path'] . '/' . $appId);
+				if (empty($versionToLoad) || version_compare($version, $versionToLoad['version'], '>')) {
```
`empty()` , fortunately we don't have a version "0" which would qualify as true for `empty("0")`

VicDeo replied:
OMG! an evil app-developer can write an app that will never be loaded :) 
Just kidding. `$versionToLoad` is an array so `empty` is valid in this context.

PVince81 reviewed (approved):
:+1: looks good to me.

Feel free to ignore the comments if they apply on old code you just moved to another location. If you do fix them, let's make sure there is no risk of regression.

VicDeo commented:
@PVince81 I changed comparison to be strict and removed double space.
`findAppInDirectories` always returns either false or non-empty array so these changes are regression-safe. (going to retest)

PVince81 commented on lib/public/App/IAppManager.php:
```diff
@@ -166,4 +166,26 @@ public function readAppPackage($path);
 	public function canInstall();
+
+	/**
+	 * Get the directory for the given app.
```
mind adding a detail whether this is an absolute path or relative path (and relative to what)

PVince81 commented on lib/public/App/IAppManager.php:
```diff
@@ -166,4 +166,26 @@ public function readAppPackage($path);
+	 * @return string|false
+	 * @since 10.0.5
+	 */
+	public function getAppWebPath($appId);
```
what's a web path ? maybe an example in PHPDoc will help clarify

PVince81 reviewed (approved):
Please check the two clarifications (optional)

VicDeo commented:
@PVince81 thanks, will do

//Just for my reference as both comments are related to 
 `apps_paths` content https://github.com/owncloud/core/blob/master/config/config.sample.php#L715

VicDeo committed:
Add more explanations to the method descriptions

VicDeo commented:
@PVince81 PHPDocs updated

PVince81 commented:
tests were green before, only PHP Doc changes, insta-merge

PVince81 commented:
@VicDeo please backport to stable10

VicDeo commented:
Stable10: https://github.com/owncloud/core/pull/30041
